### PR TITLE
Add tests for `pickle` and `cloudpickle` serializability of model definitions

### PR DESCRIPTION
**Pull request by edoakes**
## Change Summary

Adds tests for end-to-end `pickle` and `cloudpickle` (de-)serialization.

## Related issue number

Closes https://github.com/pydantic/pydantic/issues/6763

## Checklist

* [x] The pull request title is a good summary of the changes - it will be used in the changelog
* [x] Unit tests for the changes exist
* [x] Tests pass on CI
* [x] Documentation reflects the changes where applicable
* [x] My PR is ready to review, **please add a comment including the phrase "please review" to assign reviewers**

Selected Reviewer: @hramezani

### CODE REVIEW

**Comment by edoakes** [comment]
@davidhewitt PTAL

File: tests/test_pickle_pydantic_weakref.py [hunk 1]
```diff
@@ -1,52 +0,0 @@
-import gc
```
Comment (edoakes) [inline tests/test_pickle_pydantic_weakref.py:1]: These are moved to `tests/test_pickle.py`

**Comment by edoakes** [comment]
Maybe I need to use the magical incantation in the PR template: please review @davidhewitt

**Review by davidhewitt — APPROVED** [review]
Looks great to me, many thanks for adding these!

**Pull request merged**